pkg/serverstate/statetest/test_project.go

```diff
@@ -1064,3 +1067,116 @@ func TestProjectCount(t *testing.T, factory Factory, restartF RestartFactory) {
+
+	s := factory(t)
+	defer s.Close()
+
+	// Create a project template
+	projectTemplate, err := s.CreateProjectTemplate(ctx, &pb.ProjectTemplate{
+		Id:                     "",
+		Name:                   "test",
+		Summary:                "test",
+		ExpandedSummary:        "test",
+		ReadmeMarkdownTemplate: []byte("test"),
+		TerraformNocodeModule: &pb.ProjectTemplate_TerraformNocodeModule{
+			Source:  "test",
+			Version: "test",
+		},
+		Tags: []string{"test1", "test2"},
+	})
+	require.NoError(err)
+
+	name := "p1"
+
+	proj := serverptypes.TestProject(t, &pb.Project{
+		Name:           name,
+		ReadmeMarkdown: []byte("contents"),
+		ProjectTemplate: &pb.Ref_ProjectTemplate{
+			Ref: &pb.Ref_ProjectTemplate_Id{
+				Id: projectTemplate.Id,
+			},
+		},
+		FileChangeSignal: "SIGINT",
+	})
+	// Can save by ID
+	err = s.ProjectPut(ctx, proj)
+	require.NoError(err)
+
+	// Can read
+	{
+		resp, err := s.ProjectGet(ctx, &pb.Ref_Project{Project: name})
+		require.NoError(err)
+		require.Empty(cmp.Diff(proj, resp, protocmp.Transform()))
+	}
+
+	// Can update
+	{
+		proj.ReadmeMarkdown = []byte("different contents")
+
+		err := s.ProjectPut(ctx, proj)
+		require.NoError(err)
+		resp, err := s.ProjectGet(ctx, &pb.Ref_Project{Project: name})
+		require.NoError(err)
+		require.Empty(cmp.Diff(proj, resp, protocmp.Transform()))
+	}
+
+	// Cannot set a project ref to a non-existent project template
+	{
+		proj.ProjectTemplate.Ref = &pb.Ref_ProjectTemplate_Id{Id: "does-not-exist"}
+
+		err := s.ProjectPut(ctx, proj)
+		require.Error(err)
+	}
+
+	// Can set the project template ref by name
+	projectTemplate2, err := s.CreateProjectTemplate(ctx, &pb.ProjectTemplate{
+		Id:                     "",
+		Name:                   "test2",
+		Summary:                "test",
+		ExpandedSummary:        "test",
+		ReadmeMarkdownTemplate: []byte("test"),
+		TerraformNocodeModule: &pb.ProjectTemplate_TerraformNocodeModule{
+			Source:  "test",
+			Version: "test",
+		},
+		Tags: []string{"test1", "test2"},
+	})
+	require.NoError(err)
+	{
+		// Create a project template
+		proj.ProjectTemplate.Ref = &pb.Ref_ProjectTemplate_Name{Name: projectTemplate2.Name}
+		err = s.ProjectPut(ctx, proj)
+		require.NoError(err)
+
+		resp, err := s.ProjectGet(ctx, &pb.Ref_Project{Project: name})
+		require.NoError(err)
+
+		// Can come back as name or ID
+		switch ref := resp.ProjectTemplate.Ref.(type) {
+		case *pb.Ref_ProjectTemplate_Name:
+			require.Equal(projectTemplate2.Name, ref.Name)
+		case *pb.Ref_ProjectTemplate_Id:
+			require.Equal(projectTemplate2.Id, ref.Id)
+		}
+	}
+
+	// Deleting the project template doesn't cause an error when getting projects
+	{
+		//require.NoError(
+		err := s.DeleteProjectTemplate(ctx, &pb.Ref_ProjectTemplate{
+			Ref: &pb.Ref_ProjectTemplate_Id{
+				Id: projectTemplate2.Id,
+			},
+		})
+		//)
+
+		resp, err := s.ProjectGet(ctx, &pb.Ref_Project{Project: name})
+		require.NoError(err)
+
+		require.Empty(resp.ProjectTemplate)
```

Comment:
Based on this, I'm assuming that our "expected behavior" for project details where the template no longer exists is to still return the project proto, just with an empty template field.

Reply:
That should be more explicit, good catch!

Reply:
I could imagine that users would appreciate us leaving a breadcrumb here, so they could determine the project was created from a template that no longer exists.

It's tricky for other state implementations though. Imagine this as a foreign-key reference to a project template table. Once the project template record is deleted, a bare primary key on project doesn't give us much!

Looking at it today, I think deleting this require is the right call. Clearing the project template ref isn't something think we care about explicitly, and we can follow up in the future with a better breadcrumb as a separate feature.